### PR TITLE
Lido withdraw strategy

### PR DESCRIPTION
**Pull request by naddison36**
# Contract changes

* Added new `LidoWithdrawalStrategy` contract to swap stETH or WETH using the Lido Withdrawa Queue

# Dependencies

* https://github.com/OriginProtocol/origin-dollar/pull/2015

# Security

If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. 
  - [ ] Copy & paste code review [security checklist](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md) below this checklist.
  - [ ] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    /**
+     * @notice Takes all given stETH and creates Lido withdrawal request
+     */
+    function depositAll() external override onlyVault nonReentrant {
+        uint256 stETHStart = stETH.balanceOf(address(this));
+        require(stETHStart > 0, "No stETH to withdraw");
+
+        uint256 numWithdrawals = (stETHStart / MaxWithdrawalAmount) + 1;
```
Comment (pandadefi) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:137]: In case `stETHStart mod MaxWithdrawalAmount == 0` the +1 is not necessary.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        );
+
+        emit WithdrawalRequests(requestIds, amounts);
+
+        // Is there any stETH left except 1 wei from each request?
+        require(
+            stETH.balanceOf(address(this)) <= numWithdrawals,
+            "Not all stEth in withdraw queue"
+        );
+        outstandingWithdrawals += stETHStart; // Single set for gas reasons
```
Comment (pandadefi) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:160]: You could subscribe to stETHStart the very little balance of stETH.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        uint256[] memory _requestIds,
+        uint256 expectedAmount
+    ) external nonReentrant {
+        require(
+            msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Caller is not the Strategist"
+        );
+        uint256 startingBalance = payable(address(this)).balance;
+        uint256 lastIndex = withdrawalQueue.getLastCheckpointIndex();
+        uint256[] memory hintIds = withdrawalQueue.findCheckpointHints(
```
Comment (pandadefi) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:209]: Have you considered doing this off-chain and passing the hint to the function?

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        if (payable(address(this)).balance > 0) {
+            weth.deposit{ value: payable(address(this)).balance }();
+        }
+        uint256 wethBalance = weth.balanceOf(address(this));
+        if (wethBalance > 0) {
+            // slither-disable-next-line unchecked-transfer
+            weth.transfer(vaultAddress, wethBalance);
+            emit Withdrawal(address(weth), address(0), wethBalance);
+        }
+        uint256 fraxEthBalance = stETH.balanceOf(address(this));
```
Comment (pandadefi) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:256]: Here the name of the function refers to an other liquid staking protocol.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        // slither-disable-next-line unchecked-transfer
+        weth.transfer(vaultAddress, currentBalance);
+        emit Withdrawal(
+            address(weth),
+            address(withdrawalQueue),
+            currentBalance
+        );
+    }
+
+    function _abstractSetPToken(address, address) internal override {
```
Comment (DanielVF) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:238]: Maybe throw this internal method to the bottom of the contract? So it's not between all the external withdraw methods.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 1 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:137]: `numWithdrawals` is how many withdrawal requests we are sending to the Lido withdrawal queue.
When the stETH balance is less than the `MaxWithdrawalAmount` (1000 ether), then there is one withdrawal.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 3 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:209]: I like the thinking as it'll save gas on the `getLastCheckpointIndex` and `findCheckpointHints` calls. Looking at our stETH ARM, it'll save around 16k in gas.

059e3103

Given this is one-off use, I think can leave the `getLastCheckpointIndex` and `findCheckpointHints` calls in the contract. gas is relatively cheap at the moment :-)

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 4 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:256]: good pickup. I did a search and replace for frxEth but missed fraxEth

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 5 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:238]: done. I've also made it pure

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:142]: good idea. I've gone with `withdrawalLength`

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 2 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol]: Good point. I've changed it to
```Solidity
// Is there any stETH left except 1 wei from each request?
// This is because stETH does not transfer all the transfer amount.
uint256 stEthDust = stETH.balanceOf(address(this));
require(
    stEthDust <= withdrawalLength,
    "Not all stEth in withdraw queue"
);
outstandingWithdrawals += stETHStart - stEthDust;
```

**Comment by shahthepro** [comment]
Since https://github.com/OriginProtocol/origin-dollar/pull/2097 deployed 097 already and this PR seems to be update that, we need a different deployment file

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        uint256 stETHStart = stETH.balanceOf(address(this));
+        require(stETHStart > 0, "No stETH to withdraw");
+
+        uint256 withdrawalLength = (stETHStart / MaxWithdrawalAmount) + 1;
+        uint256[] memory amounts = new uint256[](withdrawalLength);
+
+        uint256 stETHRemaining = stETHStart;
+        uint256 i = 0;
+        while (stETHRemaining > MaxWithdrawalAmount) {
+            amounts[i++] = MaxWithdrawalAmount;
```
Comment (sparrowDom) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:148]: Nice I haven't seen this expression in years. Had too google that the [i++] returns the initial value but after the expression still increments it.

File: contracts/test/strategies/lido_withdrawal_strategy.fork-test.js [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        .connect(strategist)
+        .withdrawAllFromStrategy(lidoWithdrawalStrategy.address);
+
+      // Check final balances
+      const finalWethBalanceVault = await weth.balanceOf(oethVault.address);
+      const finalstEthBalanceVault = await stETH.balanceOf(oethVault.address);
+      const finalNativeEthBalanceVault = await oethVault.provider.getBalance(
+        oethVault.address
+      );
+      expect(finalWethBalanceVault.sub(initialWethBalanceVault))
```
Comment (sparrowDom) [inline contracts/test/strategies/lido_withdrawal_strategy.fork-test.js:119]: very good fork tests 🙏

**Comment by sparrowDom** [comment]
@shahthepro yeah the diff on 097 deployment file is pretty terrible :) there is a clean [098 deployment file](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/deploy/mainnet/098_lido_withdrawal_strategy.js) in master that is ok.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    IWETH9 private constant weth =
+        IWETH9(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    /// @notice Address of the stETH token
+    IERC20 private constant stETH =
+        IERC20(0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84);
+    /// @notice Address of the Lido Withdrawal Queue contract
+    IStETHWithdrawal private constant withdrawalQueue =
+        IStETHWithdrawal(0x889edC2eDab5f40e902b864aD4d7AdE8E412F9B1);
+    /// @notice Maximum amount of stETH that can be withdrawn in a single request
+    uint256 public constant MaxWithdrawalAmount = 1000 ether;
```
Comment (shahthepro) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:94]: nit: For constants, using upper snake case would be nice, not necessary to change at this point

**Comment by shahthepro** [comment]
## Requirements

The strategy should interact with Lido Withdrawal Queue and should be able to redeem stETH.

## Easy Checks

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication. 
- - _`claimWithdrawals` is callable by Strategist but not by Governor. Ideally I would expect Governor to be able to do everything that Strategist can do. However this is fine, since this is a one-off strategy and if needed the Governor can still call `setStrategist()` and then this method in a single proposal._


#### Ethereum

- [ ] Contract does not send or receive Ethereum.
- - _It does receive ETH and wrap it into WETH during withdrawals. However, that's how it's supposed to be`_
- [ ] Contract has no payable methods.
- - _It needs to receive ETH, so needs the `receive() payable` method that it has_
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [x] Contracts with for loops must have either:
    - [ ] A way to remove items
    - [ ] Can be upgraded to get unstuck
    - [ ] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- [x] All for loops use uint256

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

- - _Nit: The contract's initialize method takes three args, however we always expect those three to be empty arrays. So, could probably change the function signature but not necessary_

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding
- [x] Contract rounds in the protocols favor
- [x] Contract does not have bugs from loosing rounding precision
- [x] Code correctly multiplies before division
- [x] Contract does not have bugs from zero or near zero amounts

#### Dependencies
- [x] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes.
- [x] If OpenZeppelin ACL roles are use review & enumerate all of them.
- [x] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used.

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] No malicious behaviors
- [x] Low level call() must require success.
- [x] No slippage attacks (we need to validate expected tokens received)
- [ ] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [x] Each publicly callable method has a test
  - [x] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing.

#### Deploy

- [x] Deployer permissions are removed after deploy

## Strategy Specific

_Remove this section if the code being reviewed is not a strategy._

#### Strategy checks

- [x] Check balance cannot be manipulated up AND down by an attacker
- [x] No read only reentrancy on downstream protocols during checkBalance
- [ ] All reward tokens are collected
  - _No reward tokens_
- [ ] The harvester can sell all reward tokens
- [ ] No funds are left in the contract that should not be as a result of depositing or withdrawing
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [x] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [x] WithdrawAll() cannot be MEV'd
- [x] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances
- [ ] We have monitoring on a pauses in all downstream systems

## Thinking

#### Logic

  - _There's no `onERC721Receive` method on this contract. However, Lido contract doesn't call that method when minting. Those are called/checked only during transfers. This kinda works for us since the strategy won't receive any NFT that it didn't mint_

  - [x] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

_Deployment seems to be split and moved to a different file on master, so leaving it out of this review_

#### Internal State

_Everything seems about right. The contract depends on `outstandingWithdrawals` for all withdrawal accounting._

#### Attack

_Doesn't seem to be vulnerable to common attacks._

#### Flavor

_Code is simple and elegant. Doesn't seem to have any vulnerability_

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        uint256 withdrawalLength = (stETHStart / MaxWithdrawalAmount) + 1;
+        uint256[] memory amounts = new uint256[](withdrawalLength);
+
+        uint256 stETHRemaining = stETHStart;
+        uint256 i = 0;
+        while (stETHRemaining > MaxWithdrawalAmount) {
+            amounts[i++] = MaxWithdrawalAmount;
+            stETHRemaining -= MaxWithdrawalAmount;
+        }
+        amounts[i] = stETHRemaining;
```
Comment (sparrowDom) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:151]: Does this revert if the `stETHRemaining` == 0? Though considering current stETH balance on the vault is ` 17,886.18` this is highly unlikely to happen

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    /// @notice Total amount of stETH that has been requested to be withdrawn for ETH
+    uint256 public outstandingWithdrawals;
+
+    event WithdrawalRequests(uint256[] requestIds, uint256[] amounts);
+    event WithdrawalClaims(uint256[] requestIds, uint256 amount);
+
+    constructor(BaseStrategyConfig memory _stratConfig)
+        InitializableAbstractStrategy(_stratConfig)
+    {
+        require(MaxWithdrawalAmount < type(uint120).max);
```
Comment (sparrowDom) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:104]: we can remove this line since it will always be true. To anyone interested
```
>> type(uint120).max / 1e18
>> 1329227995784915872
```

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 9 above)
Comment (sparrowDom) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:151]: Confirmed it would revert since according to the docs (https://docs.lido.fi/contracts/withdrawal-queue-erc721/): 

_The minimal amount for a request is 100 wei, and the maximum is 1000 eth. More significant amounts should be split into several requests, which allows us to avoid clogging the queue with an extra large request._

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 8 above)
Comment (sparrowDom) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:94]: The documentation (https://docs.lido.fi/contracts/withdrawal-queue-erc721/#claimwithdrawals) states that the withdrawal amounts must be lower than the max amount: 
_each amount in _amounts must be greater than MIN_STETH_WITHDRAWAL_AMOUNT and lower than MAX_STETH_WITHDRAWAL_AMOUNT_ 

and not lower or equal

Comment (sparrowDom) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:94]: Seems the documentation is wrong. The source code https://etherscan.io/address/0xe42c659dc09109566720ea8b2de186c2be7d94d9#code is as follows: 
```
uint256 public constant MIN_STETH_WITHDRAWAL_AMOUNT = 100;
uint256 public constant MAX_STETH_WITHDRAWAL_AMOUNT = 1000 * 1e18;

function _checkWithdrawalRequestAmount(uint256 _amountOfStETH) internal pure {
        if (_amountOfStETH < MIN_STETH_WITHDRAWAL_AMOUNT) {
            revert RequestAmountTooSmall(_amountOfStETH);
        }
        if (_amountOfStETH > MAX_STETH_WITHDRAWAL_AMOUNT) {
            revert RequestAmountTooLarge(_amountOfStETH);
        }
    }
```

which obviously means withdrawal amount can be 1000 ether exactly

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * @notice Claim previously requested withdrawals that have now finalized.
+     * Called by the Strategist.
+     * @param _requestIds Array of withdrawal request identifiers
+     * @param expectedAmount Total amount of ETH expect to be withdrawn
+     */
+    function claimWithdrawals(
+        uint256[] memory _requestIds,
+        uint256 expectedAmount
```
Comment (sparrowDom) [inline contracts/contracts/strategies/LidoWithdrawalStrategy.sol:208]: super nitpicky: this could use an underscore `_expectedAmount`

**Comment by sparrowDom** [comment]
## Requirements

The strategy contract is interacting with Lido withdrawal queue to natively redeem stETH for ETH.

## Easy Checks

#### Authentication

- [x] Never use tx.origin
- [x] Every external/public function is supposed to be externally accessible
- [x] Every external/public function has the correct authentication

#### Ethereum

- 🟠  Contract does not send or receive Ethereum. _It does receive it but it performs the checks for expected ETH received ignoring the starting contract balance._ 
- 🟠 Contract has no payable methods. _It has it needs it_
- [x] Contract is not vulnerable to being sent self destruct ETH

#### Cryptographic code

- [x] This contract code does not roll it's own crypto.
- [x] No signature checks without reverting on a 0x00 result.
- [x] No signed data could be used in a replay attack, on our contract or others.

#### Gas problems

- [x] Contracts with for loops must have either: _has a while loop_
    - [ ] A way to remove items
    - [ ] Can be upgraded to get unstuck
    - [ ] Size can only controlled by admins
    - [x] loop is designed in a way that it definitely exits 
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying. _If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing._
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Overflow

- [x] Code is solidity version >= 0.8.0
- [] All for loops use uint256 _no for loops_

#### Proxy

- [x] No storage variable initialized at definition when contract used as a proxy implementation.

#### Events
- [x] All state changing functions emit events

## Medium Checks

#### Rounding
- [x] Contract rounds in the protocols favor
- [x] Contract does not have bugs from loosing rounding precision
- [] Code correctly multiplies before division _doesn't divide in a way this would be needed_
- [] Contract does not have bugs from zero or near zero amounts _It potentially does since stETHRemaining in `depositAll` could be 0. But it is highly unlikely and I wouldn't fix it_

#### Dependencies
- [ ] Review any new contract dependencies thoroughly (e.g. OpenZeppelin imports) when new dependencies are added or version of dependencies changes. _no OZ imports_
- [ ] If OpenZeppelin ACL roles are use review & enumerate all of them. _not used_
- [ ] Check OpenZeppelin [security vulnerabilities](https://github.com/OpenZeppelin/openzeppelin-contracts/security/advisories) and see if any apply to current PR considering the version of OpenZeppelin contract used. _not needed_

#### External calls

- [x] Contract addresses passed in are validated
- [x] No unsafe external calls
- [x] Reentrancy guards on all state changing functions
    - [ ] Still doesn't protect against external contracts changing the state of the world if they are called.
- [x] No malicious behaviors
- [x] Low level call() must require success. _none used_
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Do not call balanceOf for external contracts to determine what they will do when they use internal accounting

#### Tests

  - [x] Each publicly callable method has a test
  - [x] Each logical branch has a test
  - [ ] Each require() has a test _no, but not needed IMO_
  - [ ] Edge conditions are tested _this will be a 1 off called only by the strategist. Edge cases don't need covering_
  - [ ] If tests interact with AMM make sure enough edge cases (pool tilts) are tested. Ideally with fuzzing. _they don't_

#### Deploy

- [x] Deployer permissions are removed after deploy

## Strategy Specific
#### Strategy checks

- [x] Check balance cannot be manipulated up AND down by an attacker
- [x] No read only reentrancy on downstream protocols during checkBalance
- [x] All reward tokens are collected
- [ ] The harvester can sell all reward tokens _harvester not utilized_
- [x] No funds are left in the contract that should not be as a result of depositing or withdrawing
- [x] All funds can be recovered from the strategy by some combination of depositAll, withdraw, or withdrawAll()
- [x] WithdrawAll() can always withdraw an amount equal to or larger than checkBalances report, even in spite of attacker manipulation.
- [x] WithdrawAll() cannot be MEV'd
- [x] Strategist cannot steal funds

#### Downstream

- [ ] We have monitoring on all backend protocol's governances _no not needed_
- [ ] We have monitoring on a pauses in all downstream systems _no not needed_

## Thinking

#### Logic

  - [x] Correct usage of global & local variables. -> they might differentiate only by an underscore that can be overlooked (e.g. address vs _address).

#### Deployment Considerations

No this is a pretty straight forward deploy. 

#### Internal State

- What can be always said about relationships between stored state
The amount stored in `outstandingWithdrawals` always matches the amount that is being in the Lido's withdrawal queue. And only the owner of the requests can claim them, meaning no one else aside from strategy contract can claim the items in the queue and potentially manipulate the `outstandingWithdrawals`. 
- What must hold true about state before a function can run correctly (preconditions)
`outstandingWithdrawals` makes sure that checkBalance correctly represents the funds in the strategy. Any additional native redemptions or claims accurately update the `outstandingWithdrawals`.
- What must hold true about the return or any changes to state after a function has run.
Same as the point a line above. 

Does this code do that?
Yes

#### Attack

_What could the impacts of code failure in this code be._
If someone would be able to influence the amount returned in `checkBalance`. Or interfere with the Lido's withdrawal queue lifecycle. 

_What conditions could cause this code to fail if they were not true._
If Lido's withdrawal queue wouldn't be permissioned. If code in our contract would read the balance of tokens instead of keeping its own accounting that is immune to unexpected transfers

_Does this code successfully block all attacks._
yes 
#### Flavor
Code is very clean and simple.

**Comment by naddison36** [comment]
I mistakenly merged this PR to `master` when I merged the soETH changes.

I'll create a new branch for post merge changes `nicka/lido-withdraw-strategy2`.

**Comment by sparrowDom** [comment]
Yeah I know. I've left some comments in the code, though those 2 comments that are actionable are really nitpicky. I don't think you need to do any changes.

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 8 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:94]: Good thing you checked Lido's source code. Thanks

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 10 above)
Comment (naddison36) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:104]: I've made this change on the `nicka/lido-withdraw-strategy2` branch in PR https://github.com/OriginProtocol/origin-dollar/pull/2116

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 8 above)
Comment (sparrowDom) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:94]: thanks

File: contracts/contracts/strategies/LidoWithdrawalStrategy.sol (hunk 11 above)
Comment (sparrowDom) [reply contracts/contracts/strategies/LidoWithdrawalStrategy.sol:208]: thanks
